File: backend/app/schemas/realtime.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class VehiclePosition(BaseModel):
    vehicle_id: str
    route_id: str | None = None
    trip_id: str | None = None
    latitude: float
    longitude: float
    bearing: float | None = None
    speed_kmh: float | None = None
    observed_at: datetime
    status: Literal["in_transit", "stopped", "unknown"] = "unknown"


class ArrivalPrediction(BaseModel):
    stop_id: str
    route_id: str
    trip_id: str | None = None
    vehicle_id: str | None = None
    predicted_arrival: datetime
    generated_at: datetime
    uncertainty_seconds: int | None = None
    model_version: str | None = None


class RealtimeMeta(BaseModel):
    generated_at: datetime
    count: int = Field(ge=0)
    status: Literal["live", "empty", "stale"]
    stale: bool = False
    stale_after_seconds: int = Field(ge=1)


class VehiclePositionResponse(BaseModel):
    data: list[VehiclePosition]
    meta: RealtimeMeta


class ArrivalPredictionResponse(BaseModel):
    data: list[ArrivalPrediction]
    meta: RealtimeMeta

File: frontend/types/realtime.ts
export type VehicleStatus = "in_transit" | "stopped" | "unknown";

export type VehiclePosition = {
  vehicle_id: string;
  route_id: string | null;
  trip_id: string | null;
  latitude: number;
  longitude: number;
  bearing: number | null;
  speed_kmh: number | null;
  observed_at: string;
  status: VehicleStatus;
};

export type ArrivalPrediction = {
  stop_id: string;
  route_id: string;
  trip_id: string | null;
  vehicle_id: string | null;
  predicted_arrival: string;
  generated_at: string;
  uncertainty_seconds: number | null;
  model_version: string | null;
};

export type RealtimeMeta = {
  generated_at: string;
  count: number;
  status: "live" | "empty" | "stale";
  stale: boolean;
  stale_after_seconds: number;
};

export type RealtimeResponse<T> = {
  data: T[];
  meta: RealtimeMeta;
};

export type ServiceAlert = {
  id: string;
  title: string;
  description: string | null;
  url: string | null;
  cause: string | null;
  effect: string | null;
  route_ids: string[];
  stop_ids: string[];
  active_from: string | null;
  active_until: string | null;
};

export type ServiceAlertResponse = {
  data: ServiceAlert[];
  meta: {
    generated_at: string;
    count: number;
    status: "live" | "empty" | "unavailable";
    source_configured: boolean;
  };
};

export interface RealtimeProvider {
  listVehicles(routeId?: string): Promise<VehiclePosition[]>;
  listPredictions(stopId: string, routeId?: string): Promise<ArrivalPrediction[]>;
}
